### DISCUSSION TITLE
How to show white texture in filament?

### DISCUSSION THREAD

**Discussion by udayk07** [google/filament#4915] [Q&A] [answered]
While trying to render a texture ( from Android View ), it showing as light grey instead of pure white. Used filmic tonemapper and also other tone mappers ( like linear tone mapper ) also which does't work ( all cases renders as light grey ).  

`renderer.getFilamentView().setColorGrading(
                    new ColorGrading
                            .Builder()
                            .toneMapper(new ToneMapper.Filmic())
                            .build(EngineInstance.getEngine().getFilamentEngine())
            );`

Tried to map inverseTonemap color to material's basecolor but that workaround also not works and also it makes red border into yellow color.

`material.baseColor.rgb = inverseTonemapSRGB(material.baseColor.rgb);`

![sceneform_AR](https://user-images.githubusercontent.com/78372917/144277611-bcd282f5-d7bb-4299-a513-34935e5c8334.jpg)

.mat file used for texture maping:

`material {
    "name" : "View",
    "defines" : [
        "baseColor"
    ],
    "parameters" : [
        {
           "type" : "samplerExternal",
           "name" : "viewTexture"
        },
        {
            "type" : "float2",
            "name" : "offsetUv"
        },
        {
            "type" : "float",
            "name" : "alpha"
        },
    ],
    "requires" : [
        "position",
        "uv0"
    ],
    "shadingModel" : "unlit",
    "blending" : "transparent",
    "doubleSided" : true
}
fragment {
    void material(inout MaterialInputs material) {
        prepareMaterial(material);
        vec2 uv = getUV0();
        if (!gl_FrontFacing) {
          uv.x = 1.0 - uv.x;
        }
        uv.x = uv.x + materialParams.offsetUv.x * (1.0 - 2.0 * uv.x);
        uv.y = uv.y + materialParams.offsetUv.y * (1.0 - 2.0 * uv.y);
        material.baseColor = texture(materialParams_viewTexture, uv) * materialParams.alpha;
        //material.baseColor.rgb = inverseTonemapSRGB(material.baseColor.rgb);
    }
}`

Would anyone guide me to achieve pure white color texture in scene  ?

**Comment by romainguy** [comment, marked as answer]
You will always have that problem no matter what tone mapper you use. Tone mapper takes HDR data and map it to the SDR range, which means "SDR" values in your scene *have* to be mapped to a range that's < 0..1. You have three solutions:
- Change your texture/material to use HDR values so they get mapped to SDR white
- Use `inverseTonemapSRGB()` or `inverseTonemap()` (but like you saw it can cause hue shifts, which is something I want to fix; there's an issue opened for this). Make sure you use the right one, only use `inverseTonemapSRGB()` if the color values you use are not linear. In your example above, you should be using `inverseTonemap()`
- Disable tone mapping

**Reply by udayk07** [reply]
@romainguy  thanks for the suggestions.

Tried to use `inverseTonemapSRGB()` like below in material which renders red color border as yellow one as below ( renders white color correctly ).

material fragment shader code:

`    void material(inout MaterialInputs material) {
        prepareMaterial(material);
        vec2 uv = getUV0();
        if (!gl_FrontFacing) {
          uv.x = 1.0 - uv.x;
        }
        // Set offsetUv if we want to invert around an axis.
        // In front facing camera, set offsetUv.x to 1 and offsetUv.y to 0.
        uv.x = uv.x + materialParams.offsetUv.x * (1.0 - 2.0 * uv.x);
        uv.y = uv.y + materialParams.offsetUv.y * (1.0 - 2.0 * uv.y);
        material.baseColor = texture(materialParams_viewTexture, uv) * materialParams.alpha;
        material.baseColor.rgb = inverseTonemapSRGB(material.baseColor.rgb);
    }`


![Screenshot_20211202-000510_ZC_AR](https://user-images.githubusercontent.com/78372917/144294251-774b9b91-a926-48c4-ac80-3f630c04e022.jpg)



When tried to disable tone mapping by `renderer.getFilamentView().setColorGrading(null);` which does't solve the case ( it renders environement as light yellowish as below )

![Screenshot_20211201-235650_ZC_AR](https://user-images.githubusercontent.com/78372917/144294489-91a5e006-7f33-4c89-8e9c-c1e130d75074.jpg)


Have no idea on the option `Change your texture/material to use HDR values so they get mapped to SDR white` ! Is there any sample case or help doc to try this one ? 
Thanks for the help.

**Reply by romainguy** [reply]
You should be using inverseTonemap() not inverseTonemapSRGB(), and like I said there's a known issue with hue shifts in some colors. It will be quite tricky to solve though. 

Setting color grading to null means using the default color grading. Instead, use the linear tone mapper on your color grading config.

**Reply by udayk07** [reply]
@romainguy Thanks for the reply

Misread the option u mentioned earlier. As mentioned earlier, with linear and other tonemapping works for this case. Tried with `inverseTonemap` instead of inverseTonemapSRGB which also not renders as expected. 

`renderer.getFilamentView().setColorGrading(
                    new ColorGrading
                            .Builder()
                            .toneMapper(new ToneMapper.Linear())
                            .build(EngineInstance.getEngine().getFilamentEngine())
            ); `

`material.baseColor = texture(materialParams_viewTexture, uv) * materialParams.alpha;
material.baseColor.rgb = inverseTonemap(material.baseColor.rgb);`


* Sample screen with tone mapping - linear and inverseTonemap() option
![Screenshot_20211202-010031_ZC_AR](https://user-images.githubusercontent.com/78372917/144301682-e7d27a87-432a-45f9-aee2-db11a5e3610a.jpg)

**Reply by romainguy** [reply]
If you use linear tone mapping, don't use inverseTonemap() in the material.

**Reply by udayk07** [reply]
Using linear tone mapping only works ( white color renders as expected ). Thanks for the help. @romainguy
